src/test/java/pl/project13/core/ProcessHandlerTest.java

```diff
@@ -0,0 +1,193 @@
+/*
+ * This file is part of git-commit-id-plugin-core by Konrad 'ktoso' Malawski <konrad.malawski@java.pl>
+ *
+ * git-commit-id-plugin-core is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * git-commit-id-plugin-core is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with git-commit-id-plugin-core.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package pl.project13.core;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertThrows;
+import static org.junit.Assert.assertTrue;
+
+import java.io.File;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+
+import org.junit.Test;
+
+public class ProcessHandlerTest {
+  @FunctionalInterface
+  private interface TestAction {
+    void run(Process p) throws Exception;
+  }
+
+  private static final String STDOUT_LINE = "some text written to the stdout stream; line ";
+  private static final String STDERR_LINE = "some text written to the stderr stream; line ";
+
+  private static void runJavaProcess(int exitCode, int outRepeatCount, long sleepMillis, TestAction testAction) {
+    // Creates a small Java program and starts it directly from the source file (JEP 330)
+    // This is done instead of using shell scripts to be more independent of the OS running the test
+    String javaBinaryPath = System.getProperty("java.home") + File.separator + "bin" + File.separator + "java";
```

Comment:
Constructing the `javaBinaryPath` this way might be a bit brittle, but it is probably the most reliable approach.

I also thought about obtaining the `java` path using `java.lang.ProcessHandle.current()` assuming that the current test process is a Java process, but that might be too speculative. For example when running the test in Eclipse this was actually `javaw`, which is probably fine as well, but shows that making assumptions about the current running process is not reliable enough.

Any suggestions for alternative ways for obtaining the `java` path are welcome.

(Though I would really like to avoid using shell scripting instead of a Java source file, because that would require OS-specific shell scripts and might be quite error-prone.)

Reply:
Thanks for your contribution! I like the overall change and it looks good to me. However this tests feels a bit off in my view. I clearly would agree that adding tests always bring a benefit (e.g. to see if something is broken that worked before), but I'm not sure if we somehow want to rely on executing java.
I'll need to think about that...I can't think of a real alternative that would work os independant. Certainly hacking some shell scripts is IMHO even worse...

Reply:
Indeed the main problem is obtaining the java executable....

I guess you are right, in case this becomes a hassle in the future I still can remove the tests, or add @Ignore...

Thanks again for your contribution I'm going ahead and merge now!